# <font color='red'><b>Bootstrap assignment</b> </font>

<b>There will be some functions that start with the word "grader" ex: grader_sampples(), grader_30().. etc, you should not change those function definition.

Every Grader function has to return True.</b>

<font color='blue'> <b>Importing packages</b> </font>

In [1]:
import numpy as np # importing numpy for numerical computation
from sklearn.datasets import load_boston # here we are using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # importing mean_squared_error metric
from sklearn.tree import DecisionTreeRegressor
import random

In [2]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

In [3]:
x.shape

(506, 13)

In [4]:
x[:5]

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, 0.0000e+00, 5.3800e-01,
        6.5750e+00, 6.5200e+01, 4.0900e+00, 1.0000e+00, 2.9600e+02,
        1.5300e+01, 3.9690e+02, 4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        6.4210e+00, 7.8900e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9690e+02, 9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        7.1850e+00, 6.1100e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9283e+02, 4.0300e+00],
       [3.2370e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        6.9980e+00, 4.5800e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9463e+02, 2.9400e+00],
       [6.9050e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        7.1470e+00, 5.4200e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9690e+02, 5.3300e+00]])

## <font color='red'><b>Task 1</b></font>

<font color='red'> <b>Step - 1</b></font>

*  <font color='blue'><b>Creating samples</b></font><br>
    <b> Randomly create 30 samples from the whole boston data points</b>
    *  Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points
    
     For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]
* <font color='blue'><b> Create 30 samples </b></font>
    *  Note that as a part of the Bagging when you are taking the random samples <b>make sure each of the sample will have different set of columns</b><br>
Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...
Make sure each sample will have atleast 3 feautres/columns/attributes

<font color='red'><b>Step - 2 </b></font>

<font color='blue'><b>Building High Variance Models on each of the sample and finding train MSE value</b></font>

*  Build a regression trees on each of 30 samples.
*  Computed the predicted values of each data point(506 data points) in your corpus.
*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
*  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

<font color='red'> <b>Step - 3 </b></font>

*  <font color='blue'><b>Calculating the OOB score </b></font>

*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
*  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

# <font color='red'><b>Task 2</b></font>

*  <font color='blue'><b>Computing CI of OOB Score and Train MSE</b></font>
  *   Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intravels of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>
<li> Note: Refer the Central_Limit_theorem.ipynb to check how to find the confidence intravel</li>
</ol>

# <font color='red'><b>Task 3</b></font>

*  <font color='blue'><b>Given a single query point predict the price of house.</b></font>

Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
Predict the house price for this point as mentioned in the step 2 of Task 1.

<br><br><br>

# <font color='blue'> <b>Task - 1</b></font>

<font color='blue'><b>Step - 1</b></font>

*  <font color='blue'> <b>Creating samples</b></font>

<font color='Orange'><b>Algorithm</b></font>

![alt text](https://i.imgur.com/BTVYXQ1.jpg/)

*  <font color='blue'><b> Write code for generating samples</b></font>

In [5]:
def generating_samples(x,y):
    selected_rows = random.sample(range(0, 505), 303)
    selected_sample_data = x[selected_rows]
    target_selected_values = y[selected_rows]
    replicated_rows = random.sample(selected_rows,203)
    replicated_sample_data = x[replicated_rows]
    target_replicated_values = y[replicated_rows]
    no_of_selected_cols = random.randint(3,13)
    selected_cols = random.sample(range(0,13),no_of_selected_cols)
    final_selected_sample_data = selected_sample_data[:,selected_cols]
    final_replicated_sample_data = replicated_sample_data[:,selected_cols]
    final_sample_data = np.vstack((final_selected_sample_data, final_replicated_sample_data))
    final_target_values = np.vstack((target_selected_values.reshape(-1,1), target_replicated_values.reshape(-1,1)))
    return final_sample_data,final_target_values,selected_rows,selected_cols

    '''In this function, we will write code for generating 30 samples '''
    # you can use random.choice to generate random indices without replacement
    # Please have a look at this link https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html for more details
    # Please follow above pseudo code for generating samples 
    

    # return sampled_input_data , sampled_target_data,selected_rows,selected_columns
    #note please return as lists

<font color='cyan'> <b> Grader function - 1 </b> </fongt>

In [6]:
def grader_samples(a,b,c,d):
    length = (len(a)==506  and len(b)==506)
    sampled = (len(a)-len(set([str(i) for i in a]))==203)
    rows_length = (len(c)==303)
    column_length= (len(d)>=3)
    assert(length and sampled and rows_length and column_length)
    
    return True
a,b,c,d = generating_samples(x, y)
grader_samples(a,b,c,d)

True

*  <font color='blue'> <b>Create 30 samples </b>

![alt text](https://i.imgur.com/p8eZaWL.jpg)

In [7]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list

def number_of_samples(n,x,y):
    list_input_data =[]
    list_output_data =[]
    list_selected_row= []
    list_selected_columns=[]


    for i in range(0,n):
        a,b,c,d = generating_samples(x, y)
        list_input_data.append(a)
        list_output_data.append(b)
        list_selected_row.append(c)
        list_selected_columns.append(d)
    return list_input_data,list_output_data,list_selected_row,list_selected_columns

In [8]:
list_input_data,list_output_data,list_selected_row,list_selected_columns = number_of_samples(30,x,y)

<font color='cyan'> <b>Grader function - 2 </b></font>

In [9]:
def grader_30(a):
    assert(len(a)==30 and len(a[0])==506)
    return True
grader_30(list_input_data)

True

<br>

<font color='red'><b>Step - 2 </b></font>

<font color='orange'><b>Flowchart for building tree</b></font>

![alt text](https://i.imgur.com/pcXfSmp.png)

*  <font color='blue'><b> Write code for building regression trees</b></font>

In [10]:
def model_fitting(list_input_data,list_output_data):
    list_of_models = []
    for i in range(len(list_input_data)):
        model = DecisionTreeRegressor(max_depth=None)
        model.fit(list_input_data[i],list_output_data[i])
        list_of_models.append(model)
    return list_of_models

In [11]:
list_of_all_models = model_fitting(list_input_data,list_output_data)

<font color='orange'><b>Flowchart for calculating MSE </b></font>

![alt text](https://i.imgur.com/sPEE618.png)

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

*  <font color='blue'><b> Write code for calculating MSE</b></font>

In [12]:
def prediction(list_of_all_models,list_selected_columns,data_point):
    y_for_data_point =[]
    for j in range(len(list_of_all_models)):
        y_for_data_point.append(list_of_all_models[j].predict(data_point[list_selected_columns[j]].reshape(1,-1)))
    return np.median(y_for_data_point)

In [13]:
def mean_square_error(list_of_all_models,list_selected_columns,x,y):
    y_pred = []
    for i in x:
        y_pred.append(prediction(list_of_all_models,list_selected_columns,i))
    sum_of_squares =0
    for i,j in zip(y,y_pred):
        sum_of_squares += (i-j)**2
    mean_square_error = sum_of_squares/len(y)
    return mean_square_error

In [14]:
mean_square_error(list_of_all_models,list_selected_columns,x,y)

0.20069911067193677

<font color='blue'><b>Step - 3 </b></font>

<font color='orange'><b>Flowchart for calculating OOB score</b></font>

![alt text](https://i.imgur.com/95S5Mtm.png)

Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

*  <font color='blue'><b> Write code for calculating OOB score </b></font>

In [15]:
def OOB_score(list_of_all_models,list_selected_row,list_selected_columns,x,y):
    y_prediction =[]
    for i in range(len(x)):
        y_from_loop = []
        for j in range(len(list_of_all_models)):
            if i not in list_selected_row[j]:
                y_model = list_of_all_models[j].predict(x[i][list_selected_columns[j]].reshape(1,-1))
                y_from_loop.append(y_model)
        y_prediction.append(np.median(y_from_loop))
    sum_of_square =0
    for i,j in zip(y,y_prediction):
        sum_of_square += (i-j)**2
    return (sum_of_square/len(y))


In [16]:
OOB_score(list_of_all_models,list_selected_row,list_selected_columns,x,y)

12.819788372859014

# <font color='blue'><b>Task 2</b></font>

In [17]:
# Generating 35 iterative samples
iterate_mean_square =[]
iterate_oob =[]

for iteration in range(0,35):
    list_input_data,list_output_data,list_selected_row,list_selected_columns = number_of_samples(30,x,y)
    list_of_all_models = model_fitting(list_input_data,list_output_data)
    mse = mean_square_error(list_of_all_models,list_selected_columns,x,y)
    iterate_mean_square.append(mse)
    oob = OOB_score(list_of_all_models,list_selected_row,list_selected_columns,x,y)
    iterate_oob.append(oob)
  

In [18]:
len(iterate_mean_square)
len(iterate_oob)

35

In [19]:
def confidence_interval_95_percent(iterate_list):
    mean = np.mean(iterate_list)
    SD = np.std(iterate_list)
    n = len(iterate_list)
    upper_limit = mean + 2*((SD)/(np.sqrt(n)))
    lower_limit = mean - 2*((SD)/(np.sqrt(n)))
    return lower_limit,upper_limit

In [20]:
confidence_interval_95_percent(iterate_mean_square)

(0.2228721093944961, 0.34054609580971107)

In [21]:
confidence_interval_95_percent(iterate_oob)

(12.723323272874328, 13.894961426852952)

# <font color='blue'><b>Task 3</b></font>

<font color='orange'><b>Flowchart for Task 3</b></font>

<b>Hint: </b> We created 30 models by using 30 samples in TASK-1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

![alt text](https://i.imgur.com/Y5cNhQk.png)

*  <font color='blue'><b> Write code for TASK 3 </b></font>

In [22]:
xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 

In [23]:
prediction(list_of_all_models,list_selected_columns,np.array(xq))

18.7

<br><br><br>

<font color='red'><b>Write observations for task 1, task 2, task 3 indetail</b></font>
* The 95% of the mean square values are lying between 0.22 and 0.34.
* Similarly the 95% of the OOB values are lying between 12.72 and 13.89.
* The mean square values are less (varying between 0.22 and 0.34), which mean model is predicting price close to it's actual value
* The predicted price in task 3 is 18.7